src/Build/BackEnd/Components/RequestBuilder/TargetUpToDateChecker.cs

```diff
@@ -1075,6 +1085,175 @@ internal static bool IsAnyOutOfDate<T>(out DependencyAnalysisLogDetail dependenc
             return false;
         }
 
+#if FEATURE_ENUMERATION
+        internal static bool IsAnyOutOfDateWip<T>(out DependencyAnalysisLogDetail dependencyAnalysisDetailEntry, string projectDirectory, IList<T> inputs, IList<T> outputs)
```

Comment:
These two (get oldest and get newer) are pretty similar. One way to perhaps short circuit and make these more generic is instead of finding the oldest output then checking to see if there's a newer input, pick an input at random (first) and use that as a candidate for the first iteration.

```C#
  var candidateNewest = intputs[0].DateTime;
  var newestOutput = FindFileOlderThan(outputs, candidateNewest, out bool done);
  if (!done)
  {
    FindFileNewerThan(inputs, newestOutput);
  }
```
Then `FindFileNewerThan` and `FindFileOlderThan` become the same method just with a flipped check. Plus there's a possibility that you happen to get an input that was really new (git branch change) and you stop super early.